### PR TITLE
[workflow] Defer input preparation until run

### LINKED ISSUE #18221: step failure before run when ray not connected
```
from ray import workflow

@workflow.step
def one() -> int:
    return 1

@workflow.step
def add(a: int, b: int) -> int:
    return a + b

output: Workflow[int] = add.step(100, one.step())

```

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Defer input preparation so that it won't fail until run.

## Related issue number
Closes #18221

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/workflow/common.py [hunk 1]
```diff
@@ -148,10 +148,11 @@ def slugify(value: str, allow_unicode=False) -> str:
 
 
 class Workflow:
-    def __init__(self, workflow_data: WorkflowData):
+    def __init__(self, workflow_data: WorkflowData, prepare_inputs):
```
Comment (ericl) [inline python/ray/experimental/workflow/common.py:151]:
```suggestion
    def __init__(self, workflow_data: WorkflowData, prepare_inputs: bool):
```

File: python/ray/experimental/workflow/common.py [hunk 2]
```diff
@@ -148,10 +148,11 @@ def slugify(value: str, allow_unicode=False) -> str:
...
         if workflow_data.ray_options.get("num_returns", 1) > 1:
             raise ValueError("Workflow should have one return value.")
         self._data = workflow_data
+        self._prepare_inputs = prepare_inputs
```
Comment (ericl) [inline python/ray/experimental/workflow/common.py:155]:
```suggestion
        self._prepare_inputs: bool = prepare_inputs
```

**Review by ericl — APPROVED** [review]

File: python/ray/experimental/workflow/common.py (hunk 1 above)
Comment (iycheng) [reply python/ray/experimental/workflow/common.py:151]: It's actually a Callable. let me fix this.

**Comment by iycheng** [comment]
@ericl [This](https://github.com/ray-project/ray/blob/5be6bda4cf4bce9f21143289fbf3641578ba0b8d/python/ray/experimental/workflow/tests/test_variable_mutable.py#L17-L21) one will fail due to the defer preparation.
Do you think it's ok to turn this test off for now?

I feel before beta, we need to improve workflow for several aspects:
 1) virtual actor dependency management;
 2) better object ref integration between ray & workflow (right now we use ray.put but later we should use storage for this one, and that will fix this issue).

**Pull request merged**